harness/determined/pytorch/deepspeed/_deepspeed_context.py

```diff
@@ -0,0 +1,386 @@
+        if self.distributed.size > 1:
+            if self.n_gpus > 0:
+                # We launch a separate process per GPU with LOCAL_RANK set by DeepSpeed's launcher.
+                # Each process needs to bind to a unique GPU.
+                self.device = torch.device("cuda", int(cast(str, os.environ.get("LOCAL_RANK"))))
+                torch.cuda.set_device(self.device)
+            else:
+                self.device = torch.device("cpu")
+        elif self.n_gpus > 0:
+            self.device = torch.device("cuda", 0)
+        else:
+            self.device = torch.device("cpu")
+        check.is_not_none(self.device)
+
+    def to_device(self, data: pytorch._Data) -> pytorch.TorchData:
+        """Map data to the device allocated by the Determined cluster.
+
+        Since we pass an iterable over the dataloader to `train_batch` and `evaluate_batch`
+        for DeepSpeedTrial, the user is responsible for moving data to GPU if needed.  This is
+        basically a helper function to make that easier.
+        """
+        with self._record_timing("to_device", accumulate=True):
+            return pytorch.to_device(data, self.device, self._to_device_warned_types)
+
+    def is_epoch_start(self) -> bool:
+        """
+        Returns true if the current batch is the first batch of the epoch.
+
+        .. warning::
+            Not accurate for variable size epochs.
+        """
+        if self._current_batch_idx is None:
+            raise det.errors.InternalException("Training hasn't started.")
+        if self._epoch_len is None:
+            raise det.errors.InternalException("Training DataLoader uninitialized.")
+        return self._current_batch_idx % self._epoch_len == 0
+
+    def is_epoch_end(self) -> bool:
+        """
+        Returns true if the current batch is the last batch of the epoch.
+
+        .. warning::
+            Not accurate for variable size epochs.
+        """
+        if self._current_batch_idx is None:
+            raise det.errors.InternalException("Training hasn't started.")
+        if self._epoch_len is None:
+            raise det.errors.InternalException("Training DataLoader uninitialized.")
+        return self._current_batch_idx % self._epoch_len == self._epoch_len - 1
+
+    def current_train_epoch(self) -> int:
+        if self._current_batch_idx is None:
+            raise det.errors.InternalException("Training hasn't started.")
+        if self._epoch_len is None:
+            raise det.errors.InternalException("Training DataLoader uninitialized.")
+        return self._current_batch_idx // self._epoch_len
+
+    def current_train_batch(self) -> int:
+        """
+        Current global batch index
+        """
+        if self._current_batch_idx is None:
+            raise det.errors.InternalException("Training hasn't started.")
+        return self._current_batch_idx
+
+    def overwrite_deepspeed_config(self, ds_config: Union[str, Dict]) -> Dict[str, Any]:
```

Comment:
What is this ultimately for?  Is it because the deepspeed config is just too big?  Why overwrite things instead of construct them from a few parts?

Reply:
The configs can get very long.  Also, most existing deepspeed code uses a json file to specify the deepspeed config which 
makes it hard to convince users to specify a full deepspeed config in our experiment config.